exchanges/okx/okx.go

```diff
@@ -0,0 +1,4452 @@
+// path with a JSON payload (of present)
+// URL arguments must be in the request path and not as url.URL values
+func (ok *Okx) SendHTTPRequest(ctx context.Context, ep exchange.URL, f request.EndpointLimit, httpMethod, requestPath string, data, result interface{}, authenticated bool) (err error) {
+	endpoint, err := ok.API.Endpoints.GetURL(ep)
+	if err != nil {
+		return err
+	}
+	var intermediary json.RawMessage
+	newRequest := func() (*request.Item, error) {
+		utcTime := time.Now().UTC().Format(time.RFC3339)
+		payload := []byte("")
+
+		if data != nil {
+			payload, err = json.Marshal(data)
+			if err != nil {
+				return nil, err
+			}
+		}
+		path := endpoint + requestPath
+		headers := make(map[string]string)
+		headers["Content-Type"] = "application/json"
+		if authenticated {
+			var creds *account.Credentials
+			creds, err = ok.GetCredentials(ctx)
+			if err != nil {
+				return nil, err
+			}
+			signPath := "/" + okxAPIPath + requestPath
+			var hmac []byte
+			hmac, err = crypto.GetHMAC(crypto.HashSHA256,
+				[]byte(utcTime+httpMethod+signPath+string(payload)),
+				[]byte(creds.Secret))
+			if err != nil {
+				return nil, err
+			}
+			headers["OK-ACCESS-KEY"] = creds.Key
+			headers["OK-ACCESS-SIGN"] = crypto.Base64Encode(hmac)
+			headers["OK-ACCESS-TIMESTAMP"] = utcTime
+			headers["OK-ACCESS-PASSPHRASE"] = creds.ClientID
+		}
+		return &request.Item{
+			Method:        strings.ToUpper(httpMethod),
+			Path:          path,
+			Headers:       headers,
+			Body:          bytes.NewBuffer(payload),
+			Result:        &intermediary,
+			AuthRequest:   authenticated,
+			Verbose:       ok.Verbose,
+			HTTPDebugging: ok.HTTPDebugging,
+			HTTPRecording: ok.HTTPRecording,
+		}, nil
+	}
+	if result == nil {
+		return errNilResponseField
+	}
+	err = ok.SendPayload(ctx, f, newRequest)
+	if err != nil {
+		return err
+	}
+	type errCap struct {
+		Code string      `json:"code"`
+		Msg  string      `json:"msg"`
+		Data interface{} `json:"data"`
+	}
+	var errMessage errCap
+	rv := reflect.ValueOf(result)
+	if rv.Kind() != reflect.Pointer || rv.IsNil() {
+		return errInvalidResponseParam
+	}
+	errMessage.Data = result
+	err = json.Unmarshal(intermediary, &errMessage)
+	if err != nil {
+		return err
+	}
+	code, err := strconv.ParseInt(errMessage.Code, 10, 64)
+	if err == nil && code != 0 {
+		if errMessage.Msg != "" {
+			return fmt.Errorf(" error code: %d message: %s", code, errMessage.Msg)
+		}
+		err, okay := ErrorCodes[strconv.FormatInt(code, 10)]
+		if okay {
+			return err
+		}
+		return fmt.Errorf("error code: %d", code)
+	}
+	return nil
+}
+
+// Status
+
+// SystemStatusResponse retrieves the system status.
+func (ok *Okx) SystemStatusResponse(ctx context.Context, state string) ([]SystemStatusResponse, error) {
+	params := url.Values{}
+	if state == "scheduled" || state == "ongoing" || state == "pre_open" || state == "completed" || state == "canceled" {
+		params.Set("state", state)
```

Comment:
Glorious code has suggested before to just let server side fail if incorrect state is inputted.